### DISCUSSION TITLE
How to use blitters without disturbing other planes

### DISCUSSION THREAD

**Discussion by codebroker40** [dankamongmen/notcurses#1450] [Q&A]
I wrote a simple clock program that can have a background image. I use a Visual and blit it with NCBLITT_1x1, and it looks like this:  

![Mar23::185211](https://user-images.githubusercontent.com/16948412/112229243-fac87400-8c08-11eb-8518-91639a550391.png)

Fine, but if I use another blitter, such as NCBLITT_2x1 or 2x2, it looks like this: 

![Mar23::185417](https://user-images.githubusercontent.com/16948412/112229342-3105f380-8c09-11eb-84de-a5a3d49e255b.png)

The clock area  is actually 5 planes sitting on one, and to move the whole clock I just move the one plane it sits on. I don't know why that center plane is being rendered like that. What's going on here, and how do I fix it if it's possible?

Here's the repo, if you wanna dig around: https://github.com/ColeW-Picaro/notclock

**Comment by dankamongmen** [comment]


**Reply by codebroker40** [reply]
Awesome. Terminal is alacritty, TERM is  `alacritty`.

**Comment by dankamongmen** [comment]
Ahhh, I see what's going on. Note that the cruft is coming from places where you're not emitting a glyph in your clock. You set the fg and bg of your plane's *output* to opaque (shouldn't be necessary, btw -- I assume you did that thinking it would address this issue? logical, but sadly incorrect), but that's only going to affect the places where you write. Those are properly opaque. The other cells are not being marked opaque, but are instead the default transparent and default background/foreground, with no EGC.

The solver thus gets { default background, default foreground, blitted EGC } for each cell, which looks like black and white squares.

![2021-03-24-041015_938x1417_scrot](https://user-images.githubusercontent.com/143473/112276314-f9be3380-8c56-11eb-9a7c-28bf9b400361.png)

```
[schwarzgerat](130) $ git diff
diff --git src/Clock.cpp src/Clock.cpp
index 2003021..f8f1394 100644
--- src/Clock.cpp
+++ src/Clock.cpp
@@ -59,6 +59,10 @@ initClockPlane () {
          .rows = CLOCK_ROWS,
          .cols = CLOCK_COLS };
     g_clockPlane = new ncpp::Plane (stdplane, clockOpts);
+        uint64_t chan = 0;
+        channels_set_fg_alpha(&chan, CELL_ALPHA_OPAQUE);
+        channels_set_bg_alpha(&chan, CELL_ALPHA_OPAQUE);
+        g_clockPlane->set_base (" ", 0, chan);
     g_clockPlane->set_bg_alpha (CELL_ALPHA_OPAQUE);
     g_clockPlane->set_fg_rgb (0x000000);
 }
@@ -82,6 +86,10 @@ initNumPlanes () {
         g_planes->emplace_back (new ncpp::Plane (g_clockPlane, pOpts));
         (*g_planes)[i]->set_bg_alpha (CELL_ALPHA_OPAQUE);
         (*g_planes)[i]->set_fg_alpha (CELL_ALPHA_OPAQUE);
+        uint64_t chan = 0;
+        channels_set_fg_alpha(&chan, CELL_ALPHA_OPAQUE);
+        channels_set_bg_alpha(&chan, CELL_ALPHA_OPAQUE);
+        (*g_planes)[i]->set_base (" ", 0, chan);
         if (i == 1) {
             pOpts.x += NUM_COLS + COLON_COLS;
         } else {
[schwarzgerat](0) $ 
```

i was a bit C-ish here; adapt it to idiomatic C++ as you see fit.

check out `notcurses_render(3)` for more info. i know this is kind of unintuitive, but it's the only way i could figure out to provide maximum flexibility ("three channels of transparency").

thanks for using Notcurses!

**Reply by dankamongmen** [reply]
btw: as for why it "worked" with `NCBLIT_1x1`, can you see why now? think about it for a second....

.
.
.
.
.
.


because 1x1 blits entirely with spaces, using only background color as its output. the colors were already solved; the spaces were exactly what you wanted =].

**Reply by dankamongmen** [reply]
btw if you're just doing this as an example or whatnot, you can punch it up pretty easily by using a video and `stream()`. instead of whatever timer loop was going to drive your clock, just `stream()` your media, and check there for whether the clock needs be updated. you won't need any other code changes =].

**Reply by dankamongmen** [reply]
one last thing: if you just leave the blitter blank (or explicitly choose `NCBLIT_DEFAULT`), you'll get `NCBLIT_3x2` if it's supported by your terminal (and Notcurses knows as much)). note that this will require use of `NCSCALE_STRETCH`, `NCSCALE_SCALEHI`, or `NCSCALE_HIRES`, but you're already using `NCSCALE_STRETCH`, so it'll work automatically for you (`NCBLIT_DEFAULT` otherwise picks `NCBLIT_2x1` under the assumption you want to maintain aspect ratio).

**Reply by codebroker40** [reply]
Thanks for the explanation! Makes sense.
